src/uu/tr/src/operation.rs

```diff
@@ -0,0 +1,529 @@
+                Sequence::parse_blank,
+                Sequence::parse_control,
+                Sequence::parse_digit,
+                Sequence::parse_graph,
+                Sequence::parse_lower,
+                Sequence::parse_print,
+                Sequence::parse_punct,
+                Sequence::parse_space,
+                Sequence::parse_space,
+                Sequence::parse_upper,
+                Sequence::parse_xdigit,
+                Sequence::parse_char_equal,
+                Sequence::parse_char,
+            )),
+        )))(input)
+        .map(|(_, r)| r)
+        .unwrap()
+    }
+
+    pub fn dissolve(self) -> Vec<char> {
+        match self {
+            Sequence::Char(c) => vec![c],
+            Sequence::CharRange(r) => r,
+        }
+    }
+
+    /// Sequence parsers
+
+    fn parse_char(input: &str) -> IResult<&str, Sequence> {
+        take(1usize)(input).map(|(l, r)| (l, Sequence::Char(r.chars().next().unwrap())))
+    }
+
+    fn parse_unrecognized_backslash(input: &str) -> IResult<&str, Sequence> {
+        tuple((tag("\\"), none_of("01234567")))(input).map(|(l, (_, a))| {
+            let c = match a {
+                'a' => Sequence::Char('\u{0007}'),
+                'b' => Sequence::Char('\u{0008}'),
+                'f' => Sequence::Char('\u{000C}'),
+                'n' => Sequence::Char('\u{000A}'),
+                'r' => Sequence::Char('\u{000D}'),
+                't' => Sequence::Char('\u{0009}'),
+                'v' => Sequence::Char('\u{000B}'),
+                _ => Sequence::Char(a),
+            };
+            (l, c)
+        })
+    }
+
+    fn parse_1_octal(input: &str) -> IResult<&str, Sequence> {
+        tuple((tag("\\"), one_of("01234567")))(input).map(|(l, (_, a))| {
+            (
+                l,
+                Sequence::Char(std::char::from_u32(a.to_digit(8).unwrap()).unwrap()),
+            )
+        })
+    }
+
+    fn parse_2_octal(input: &str) -> IResult<&str, Sequence> {
+        tuple((tag("\\"), one_of("01234567"), one_of("01234567")))(input).map(|(l, (_, a, b))| {
+            (
+                l,
+                Sequence::Char(
+                    std::char::from_u32(a.to_digit(8).unwrap() * 8 + b.to_digit(8).unwrap())
+                        .unwrap(),
+                ),
+            )
+        })
+    }
+
+    fn parse_3_octal(input: &str) -> IResult<&str, Sequence> {
+        tuple((
+            tag("\\"),
+            one_of("01234567"),
+            one_of("01234567"),
+            one_of("01234567"),
+        ))(input)
+        .map(|(l, (_, a, b, c))| {
+            (
+                l,
+                Sequence::Char(
+                    // SAFETY: All the values from \000 to \777 is valid based on a test below...
+                    std::char::from_u32(
+                        a.to_digit(8).unwrap() * 8 * 8
+                            + b.to_digit(8).unwrap() * 8
+                            + c.to_digit(8).unwrap(),
+                    )
```

Comment:
[`from_str_radix`](https://doc.rust-lang.org/std/primitive.u32.html#method.from_str_radix) might be helpful here. For reference, I did this once to parse hex digits:
```rust
map_opt(hex_digit1, |out: &str| {
    u32::from_str_radix(out, 16)
        .ok()
        .and_then(|u| char::from_u32(u))
})(i)
```
I think something similar for 1-3 octal digits would be:
```rust
map_opt(
    preceded(tag("\\"), recognize(many_m_n(1, 3, one_of("01234567")))),
    |out: &str| {
        u32::from_str_radix(out, 8)
            .ok()
            .and_then(|u| char::from_u32(u))
    }
)(input)
```

Reply:
Applied suggestion here. Thanks!